File: src/docker.erl
-module(docker).

-export([
    get_containers/0,
    get_images/0,
    create_container/1
]).

-record(opt, {
    name :: binary(),
    map_fun :: undefined | function(),
    is_optional :: boolean()
}).

%% Image
%% Hostname
%% Domainname
%% User
%% AttachStdin
%% AttachStdout
%% AttachStderr
%% ExposedPorts
%% Tty
%% OpenStdin
%% StdinOnce
%% Env
%% Cmd
%% Healthcheck
%% ArgsEscaped
%% Volumes
%% WorkingDir
%% Entrypoint
%% NetworkDisabled
%% MacAddress
%% OnBuild
%% Labels
%% StopSignal
%% StopTimeout
%% Shell
%% HostConfig
%% NetworkingConfig
-type optname() :: binary().

-spec get_containers() -> {ok, {StatusLine :: tuple(), Headers :: list(), DecodedJSON :: any()}} | {error, Reason :: any()}.
get_containers() ->
    json_request(get, "/containers/json", []).

-spec get_images() -> {ok, {StatusLine :: tuple(), Headers :: list(), DecodedJSON :: any()}} | {error, Reason :: any()}.
get_images() ->
    json_request(get, "/images/json", []).

-spec create_container(list({optname(), iolist()})) -> {ok, {StatusLine :: tuple(), Headers :: list(), DecodedJSON :: any()}} | {error, Reason :: any()}.
create_container(Opts) ->
    Ret = 
        parse_opts(Opts, [
            #opt{name = <<"Image">>,            is_optional = false},
            #opt{name = <<"Hostname">>,         is_optional = true},
            #opt{name = <<"Domainname">>,       is_optional = true},
            #opt{name = <<"User">>,             is_optional = true},
            #opt{name = <<"AttachStdin">>,      is_optional = true},
            #opt{name = <<"AttachStdout">>,     is_optional = true},
            #opt{name = <<"AttachStderr">>,     is_optional = true},
            #opt{name = <<"ExposedPorts">>,     is_optional = true},
            #opt{name = <<"Tty">>,              is_optional = true},
            #opt{name = <<"OpenStdin">>,        is_optional = true},
            #opt{name = <<"StdinOnce">>,        is_optional = true},
            #opt{name = <<"Env">>,              is_optional = true},
            #opt{name = <<"Cmd">>,              is_optional = true},
            #opt{name = <<"Healthcheck">>,      is_optional = true},
            #opt{name = <<"ArgsEscaped">>,      is_optional = true},
            #opt{name = <<"Volumes">>,          is_optional = true},
            #opt{name = <<"WorkingDir">>,       is_optional = true},
            #opt{name = <<"Entrypoint">>,       is_optional = true},
            #opt{name = <<"NetworkDisabled">>,  is_optional = true},
            #opt{name = <<"MacAddress">>,       is_optional = true},
            #opt{name = <<"OnBuild">>,          is_optional = true},
            #opt{name = <<"Labels">>,           is_optional = true},
            #opt{name = <<"StopSignal">>,       is_optional = true},
            #opt{name = <<"StopTimeout">>,      is_optional = true},
            #opt{name = <<"Shell">>,            is_optional = true},
            #opt{name = <<"HostConfig">>,       is_optional = true},
            #opt{name = <<"NetworkingConfig">>, is_optional = true}
        ]),
    case Ret of
        {ok, Props} ->
            json_request(post, "/containers/create", [], "application/json", jsx:encode(Props));
        {error, _} ->
            Ret
    end.

%% INTERNAL

json_request(Method, Endpoint, ReqHeaders) ->
    case docker_sock:request(Method, Endpoint, ReqHeaders) of
        {ok, {StatusLine, RespHeaders, Data}} ->
            {ok, {StatusLine, RespHeaders, jsx:decode(Data, [return_maps])}};
        Err = {error, _} ->
            Err
    end.

json_request(Method, Endpoint, ReqHeaders, ContentType, Body) ->
    case docker_sock:request(Method, Endpoint, ReqHeaders, ContentType, Body) of
        {ok, {StatusLine, RespHeaders, Data}} ->
            {ok, {StatusLine, RespHeaders, jsx:decode(Data, [return_maps])}};
        Err = {error, _} ->
            Err
    end.

parse_opts(_Opts, []) ->
    {ok, []};
parse_opts(Opts, [Opt|Tail]) ->
    case parse_opts(Opts, Tail) of
        {ok, Acc} ->
            parse_opts_(Opts, Opt, Acc);
        Err = {error, _} ->
            Err
    end.

parse_opts_(Opts, #opt{name = Name, map_fun = MapFun, is_optional = IsOptional}, Acc) ->
    case lists:keyfind(Name, 1, Opts) of
        false when not IsOptional ->
            {error, badopts};
        false ->
            {ok, Acc};
        {Name, Value0} ->
            Value1 =
                case MapFun of
                    undefined ->
                        Value0;
                    _ ->
                        MapFun(Value0)
                end,
            {ok, [{Name, iolist_to_binary(Value1)}|Acc]}
    end.

File: priv/www/js/docker.js
export class Docker {
    constructor() {
        this.fetch = async function(url, opts) {
            console.log("Fetching '%O' with opts %O", url, opts);
            let resp = await window.fetch(url, opts);
            let text = await resp.text();
            return {
                ok: resp.ok,
                status: resp.status,
                statusText: resp.statusText,
                headers: resp.headers,
                data: text.length === 0 ? null : JSON.parse(text),
            };
        }
    }

    getImage(imageId) {
        return this.fetch("/api/image/" + imageId);
    }

    getContainer(containerName) {
        if(containerName === undefined)
            return new Promise((_, reject) => reject(new Error("Invalid arguments provided! " + arguments.toNiceString())) );
        return this.fetch(`/api/socket/direct/containers/${containerName}/json`)
    }

    createContainer(containerName, imageName, command, opts) {
        opts = opts ?? {};
        if([containerName, imageName, command].any(elem => elem === undefined) || !/\/?[a-zA-Z0-9][a-zA-Z0-9_.-]+/.test(containerName))
            return new Promise((_, reject) => reject(new Error("Invalid arguments provided! " + arguments.toNiceString())) );
        console.log("containerName: %O", containerName);
        const keys = [
            'Hostname', 
            'Domainname', 
            'User', 
            'AttachStdin', 
            'AttachStdout', 
            'AttachStderr', 
            'ExposedPorts', 
            'Tty', 
            'OpenStdin', 
            'StdinOnce', 
            'Env', 
            'Healthcheck', 
            'ArgsEscaped', 
            'Volumes', 
            'WorkingDir', 
            'Entrypoint', 
            'NetworkDisabled', 
            'MacAddress', 
            'OnBuild', 
            'Labels', 
            'StopSignal', 
            'StopTimeout', 
            'Shell', 
            'HostConfig', 
            'NetworkingConfig'
        ];
        let body = {
            Cmd: command,
            Image: imageName
        };
        keys.forEach(key => {
            if(opts[key] !== undefined) 
                body[key] = opts[key];
        });
        return this.fetch(`/api/socket/direct/containers/create?name=${containerName}`, {
            method: 'POST',
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify(body),
        })
    }

    startContainer(containerName, opts) {
        opts = opts ?? {};
        if(containerName === undefined)
            return new Promise((_, reject) => reject(new Error("Invalid arguments provided! " + arguments.toNiceString())) );
        let queryParams = opts["detachKeys"] === undefined ? "" : "detachKeys=" + opts["detachKeys"];
        return this.fetch(`/api/socket/direct/containers/${containerName}/start?${queryParams}`, {
            method: 'POST'
        })
    }

    stopContainer(containerName, opts) {
        opts = opts ?? {};
        if(containerName === undefined)
            return new Promise((_, reject) => reject(new Error("Invalid arguments provided! " + arguments.toNiceString())) );
        let queryParams = opts["t"] === undefined ? "" : "t=" + opts["t"];
        return this.fetch(`/api/socket/direct/containers/${containerName}/stop?${queryParams}`, {
            method: 'POST'
        })
    }
}

window.docker = new Docker();